File: tests/fixtures/vulnerable/append-assign.py
# `+=` EXTENDS A VALUE. FOR A LONG TIME IT REPLACED IT.
#
# See append-assign.js for the full story. Python is one of the seven languages
# that lost the value, and it spells the compound assignment as its own node
# type - augmented_assignment - where Java and Go reuse the plain assignment
# node with a different operator. The fix reads the operator rather than the
# node type, so all of them are covered by one line; these fixtures are what
# says so out loud.
import os
import sys


def trailing_constant():
    cmd = "ls "
    cmd += sys.argv[1]
    cmd += " -l"
    # EXPECT-FLOW command-injection
    os.system(cmd)


def leading_constant():
    cmd = "grep "
    cmd += "-F "
    cmd += sys.argv[1]
    # EXPECT-FLOW command-injection
    os.system(cmd)


# The over-fix guard. A real overwrite must still clear the value. Reported
# here means clean assignment stopped working, which would be far worse than
# the bug this fixture exists for.
def real_reassignment_still_clears():
    cmd = "ls "
    cmd += sys.argv[1]
    cmd = "ls -l"
    os.system(cmd)

File: tests/fixtures/vulnerable/append-assign.go
// `+=` EXTENDS A VALUE. FOR A LONG TIME IT REPLACED IT.
//
// See append-assign.js for the full story. Go is worth its own fixture because
// it spells compound assignment as assignment_statement with a `+=` operator -
// the same node it uses for a plain `=`, unlike JavaScript and Python which
// have a distinct augmented node. The fix reads the operator field rather than
// the node type precisely so that this difference stops mattering, and this
// file is what proves it stopped mattering.
package main

import (
	"net/http"
	"os/exec"
)

func trailingConstant(r *http.Request) {
	cmd := "ls "
	cmd += r.URL.Query().Get("dir")
	cmd += " -l"
	// EXPECT-FLOW command-injection
	exec.Command("sh", "-c", cmd).Run()
}

func leadingConstant(r *http.Request) {
	cmd := "grep "
	cmd += "-F "
	cmd += r.URL.Query().Get("term")
	// EXPECT-FLOW command-injection
	exec.Command("sh", "-c", cmd).Run()
}

// The over-fix guard: a real overwrite must still clear the value.
func realReassignmentStillClears(r *http.Request) {
	cmd := "ls "
	cmd += r.URL.Query().Get("dir")
	cmd = "ls -l"
	exec.Command("sh", "-c", cmd).Run()
}
